kqueen/engines/openstack_kubespray.py

```diff
@@ -0,0 +1,363 @@
+                "label": "Comma separated list of nameservers",
+                "default": "1.0.0.1,8.8.8.8"
+            },
+            "availability_zone": {
+                "type": "text",
+                "order": 80,
+                "label": "Availability zone",
+                "default": "nova"
+            },
+            # TODO: add availability zone
+            # TODO: add ssh username
+        },
+        "provisioner": {
+            "auth_url": {
+                "type": "text",
+                "label": "Auth URL",
+                "order": 10,
+                "validators": {
+                    "required": True,
+                    "url": True,
+                },
+            },
+            "username": {
+                "type": "text",
+                "label": "Username",
+                "order": 20,
+                "validators": {
+                    "required": True,
+                },
+            },
+            "password": {
+                "type": "password",
+                "label": "Password",
+                "order": 30,
+                "validators": {
+                    "required": True,
+                },
+            },
+            "domain_name": {
+                "type": "text",
+                "label": "Domain name",
+                "order": 40,
+                "validators": {
+                    "required": True,
+                },
+            },
+            "project_id": {
+                "type": "text",
+                "label": "Project ID",
+                "order": 50,
+                "validators": {
+                    "required": True,
+                },
+            },
+        }
+    }
+
+    def __init__(self, cluster, *args, **kwargs):
+        super().__init__(cluster, *args, **kwargs)
+        self.stack_name = "kqueen-stack-" + cluster.id
+        self.heat = Heat(
+            auth_url=kwargs["auth_url"],
+            project_id=kwargs["project_id"],
+            username=kwargs["username"],
+            compute_api_version="3",
+            domain_name=kwargs["domain_name"],
+            identity_interface="public",
+            password=kwargs["password"],
+        )
+
+    def provision(self, *args, **kwargs):
+        app.executor.submit(self._run_provisioning, **kwargs)
+        return True, None
+
+    def _run_provisioning(self, **kwargs):
+        try:
+            stack = self.heat.create(
+                stack_name=self.stack_name,
+                package_name=HEAT_TEMPLATE_PACKAGE_NAME,
+                wait=True,
+                **kwargs)
+            # TODO: expand kwargs
+            # TODO: inject additional ssh key to download kubeconfig
+            outputs = self.heat.get_outputs(stack)
+            inventory_path = self._generate_ansible_files(outputs)
+            self._wait_for_ping(inventory_path)
+            self._run_ansible(inventory_path)
+        except Exception:
+            logger.exception("Failed to provision cluster:")
```

Comment:
why we have this structure for provision method?
imo its not good from common-base-engine arch side